craft_application/application.py

```diff
@@ -137,27 +148,42 @@ def project(self) -> models.Project:
         project_file = (self._work_dir / f"{self.app.name}.yaml").resolve()
         return self.app.ProjectClass.from_yaml_file(project_file)
 
-    def run_managed(self) -> None:
+    def run_managed(self, build_for: str | None) -> None:
         """Run the application in a managed instance."""
-        craft_cli.emit.debug(f"Running {self.app.name} in a managed instance...")
-        instance_path = pathlib.PosixPath("/root/project")
-        with self.services.provider.instance(
-            self.project.effective_base, work_dir=self._work_dir
-        ) as instance:
-            try:
-                # Pyright doesn't fully understand craft_providers's CompletedProcess.
-                instance.execute_run(  # pyright: ignore[reportUnknownMemberType,reportUnknownVariableType]
-                    [self.app.name, *sys.argv[1:]], cwd=instance_path, check=True
-                )
-            except subprocess.CalledProcessError as exc:
-                raise craft_providers.ProviderError(
-                    f"Failed to execute {self.app.name} in instance."
-                ) from exc
+        build_plan = self.project.get_build_plan()
+        extra_args: dict[str, Any] = {}
+
+        # filter out builds not matching argument `--build_for` or env `CRAFT_BUILD_FOR`
+        if build_for:
+            build_plan = [b for b in build_plan if b.build_for == build_for]
+            extra_args["env"] = {"CRAFT_BUILD_FOR": build_for}
+
+        for build_info in build_plan:
+            craft_cli.emit.debug(
+                f"Running {self.app.name} in {build_info.build_for} instance..."
+            )
+            instance_path = pathlib.PosixPath("/root/project")
+
+            with self.services.provider.instance(
+                build_info, work_dir=self._work_dir
+            ) as instance:
+                try:
+                    # Pyright doesn't fully understand craft_providers's CompletedProcess.
+                    instance.execute_run(  # pyright: ignore[reportUnknownMemberType,reportUnknownVariableType]
+                        [self.app.name, *sys.argv[1:]],
```

Comment:
`build-for` requires this to be smarter now. The command classes have a `get_managed_cmd` method that we'll have to use now, which will require passing the command into `run_managed` too.

I've also made a separate issue for handling this better than my initial design did: https://github.com/canonical/craft-application/issues/71